api/client/client.go

```diff
@@ -2170,3 +2173,79 @@ func (c *Client) GenerateCertAuthorityCRL(ctx context.Context, req *proto.CertAu
 	resp, err := c.grpc.GenerateCertAuthorityCRL(ctx, req)
 	return resp, trail.FromGRPC(err)
 }
+
+// ListResources returns a paginated list of nodes that the user has access to.
+// nextKey can be used as startKey in another call to ListResources to retrieve
+// the next page.
+// It will return a `trace.LimitExceeded` error if the page exceeds gRPC max
+// message size.
+func (c *Client) ListResources(ctx context.Context, req proto.ListResourcesRequest) ([]types.Resource, string, error) {
+	if req.Namespace == "" {
+		return nil, "", trace.BadParameter("missing parameter namespace")
+	}
+	if req.Limit <= 0 {
+		return nil, "", trace.BadParameter("nonpositive parameter limit")
+	}
+
+	resp, err := c.grpc.ListResources(ctx, &req, c.callOpts...)
+	if err != nil {
+		return nil, "", trail.FromGRPC(err)
+	}
+
+	resources := make([]types.Resource, len(resp.GetResources()))
+	for i, respResource := range resp.GetResources() {
+		var resource types.Resource
+		switch req.ResourceType {
+		case proto.ResourceType_RESOURCE_TYPE_DATABASE_SERVER:
+			resource = respResource.GetDatabaseServer()
+		case proto.ResourceType_RESOURCE_TYPE_APPLICATION_SERVER:
+			resource = respResource.GetApplicationServer()
+		default:
+			return nil, "", trace.Errorf("unsupported resource type %s", req.ResourceType)
+		}
+
+		resources[i] = resource
+	}
+
+	return resources, resp.NextKey, nil
+}
+
+// ListAllResources retrieves all the resources of a type by iterating through
+// all ListResources to pages until all the resources are listed.
+func (c *Client) ListAllResources(ctx context.Context, namespace string, resourceType proto.ResourceType) ([]types.Resource, error) {
```

Comment:
I think having `ListAllResources` and `ListResources` may be a little confusing. How about naming this function `ListResources` and the other one `ListResourcesPage` or something?